sympy/physics/mechanics/_pathway.py

```diff
@@ -0,0 +1,248 @@
+            )
+            raise TypeError(msg)
+        if len(attachments) != 2:
+            msg = (
+                f'Value {repr(attachments)} passed to `attachments` was an '
+                f'iterable of length {len(attachments)}, must be an iterable '
+                f'of length 2.'
+            )
+            raise ValueError(msg)
+        for i, point in enumerate(attachments):
+            if not isinstance(point, Point):
+                msg = (
+                    f'Value {repr(point)} passed to `attachments` at index '
+                    f'{i} was of type {type(point)}, must be {Point}.'
+                )
+                raise TypeError(msg)
+        self._attachments = tuple(attachments)
+
+    @property
+    @abstractmethod
+    def length(self) -> Expr:
+        """An expression representing the pathway's length."""
+        pass
+
+    @property
+    @abstractmethod
+    def shortening_velocity(self) -> Expr:
+        """An expression representing the pathway's shortening velocity."""
+        pass
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}(attachments={self.attachments})'
+
+
+class LinearPathway(PathwayBase):
+    """Linear pathway between a pair of attachment points.
+
+    Explanation
+    ===========
+
+    A linear pathway forms a straight-line segment between two points and is
+    the simplest pathway that can be formed. It will not interact with any
+    other objects in the system, i.e. a ``LinearPathway`` will interest other
+    objects to ensure that the path between its two ends (its attachments) is
+    the shortest possible.
+
+    Examples
+    ========
+
+    As the ``_pathway.py`` module is experimental, it is not yet part of the
+    ``sympy.physics.mechanics`` namespace. ``LinearPathway`` must therefore be
+    imported directly from the ``sympy.physics.mechanics._pathway`` module.
+
+    >>> from sympy.physics.mechanics._pathway import LinearPathway
+
+    To construct a pathway, two points are required to be passed to the
+    ``attachments`` parameter as a ``tuple``.
+
+    >>> from sympy.physics.mechanics import Point
+    >>> attachments = (Point('pO'), Point('pI'))
+    >>> linear_pathway = LinearPathway(attachments)
+    >>> linear_pathway
+    LinearPathway(attachments=(pO, pI))
+
+    The pathway create above isn't very interesting without the positions and
+    velocities of its attachment points being described. Without this its not
+    possible to describe how the pathway moves, i.e. its length or its
+    shortening velocity.
+
+    >>> from sympy.physics.mechanics import ReferenceFrame
+    >>> from sympy.physics.vector import dynamicsymbols
+    >>> N = ReferenceFrame('N')
+    >>> q = dynamicsymbols('q')
+    >>> attachments[-1].set_pos(attachments[0], q * N.x)
+    >>> attachments[-1].pos_from(attachments[0])
+    q(t)*N.x
+
+    A pathway's length can be accessed via its ``length`` attribute.
+
+    >>> linear_pathway.length
+    sqrt(q(t)**2)
```

Comment:
Note this should return `Abs(q(t))`, see https://github.com/sympy/sympy/issues/23173 (I still don't have a good solution and it shouldn't matter when moving to code gen because we pass in real floats which evaluate properly.

Reply:
@TJStienstra previously did some work looking into different options for calculating the magnitude that result in the absolute expression. He benchmarked a few options and the best candidate is below.

```python
def _magnitude(vector: Vector) -> Expr:
    """Fast magnitude that replaces ``sqrt(<...>**2)`` with ``Abs(<...>)``.

    Notes
    =====

    See https://github.com/sympy/sympy/issues/23173 for discussion about
    simplification of e.g. ``sqrt(q(t)**2)`` to ``Abs(q(t))``.

    """
    out = S.Zero
    for i, v1 in enumerate(vector.args):
        out += (v1[0].T * v1[0])[0]
        for v2 in vector.args[i + 1:]:
            out += 2 * ((v2[0].T) * v2[1].dcm(v1[1]) * (v1[0]))[0]
    if isinstance(out, Pow) and out.args[1] == 2:
        return abs(out.args[0])
    return sqrt(out)
```

Reply:
@TJStienstra can you add this to that issue for record keeping.